KAFKA-9498; Topic validation during the topic creation triggers unnecessary TopicChange events

dajac opened the pull request:
This PR avoids generating unnecessary TopicChange events during the topic validation. It does so by adding a `registerWatch` field in the `GetChildrenRequest` request. This allows to not register the watch when topics are queried from the topic validation logic.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dajac committed:
KAFKA-9498; Topic validation during the creation trigger unnecessary TopicChange events

hachikuji commented:
ok to test

ijuma commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -457,18 +457,25 @@ class KafkaZkClient private[zk] (zooKeeperClient: ZooKeeperClient, isSecure: Boo
+   * Gets all topics in the cluster. It does not register a watch.
+   * @return sequence of topics in the cluster.
+   */
+  def getAllTopicsInCluster(): Set[String] = getAllTopicsInCluster(false)
```
Overloads like this are generally not done in Scala since you can use default arguments. But maybe it's best to explicit here.

dajac replied:
Right. I have kept this method one because it is called from Java as well where default arguments do not work.

ijuma replied:
What Java code?

dajac replied:
It is used in `org.apache.kafka.streams.integration.utils.EmbeddedKafkaCluster`.

dajac replied:
We could use the other method there and pass `false` but I felt that it is more convenient like this.

dajac replied:
I have removed it finally. It is cleaner.

ijuma commented on core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala:
```diff
@@ -497,45 +497,61 @@ sealed trait AsyncRequest {
   type Response <: AsyncResponse
   def path: String
   def ctx: Option[Any]
+  def registerWatch: Boolean
```
This seems unnecessary given where we use this a few lines above.

dajac replied:
I agree. I have removed it.

ijuma reviewed (commented):
Are there places where we call this method where we only care about one topic?

dajac committed:
simplify implementation

dajac commented:
> Are there places where we call this method where we only care about one topic?

No.

dajac commented:
@ijuma or @hachikuji Do you know why we verify that the topic does not collide (`Topic.hasCollision`) with others during the topic validation? This check is done for every topic when a `CreateTopicPolicy` is used and it sounds quite expensive when the cluster has thousands of topics. It this is really necessary, what about using the metadata cache? The guarantee is not strong anyway because a colliding topic could be created between the verification and the insertion.

```scala
    if (zkClient.topicExists(topic))
      throw new TopicExistsException(s"Topic '$topic' already exists.")
    else if (Topic.hasCollisionChars(topic)) {
      val allTopics = zkClient.getAllTopicsInCluster
      // check again in case the topic was created in the meantime, otherwise the
      // topic could potentially collide with itself
      if (allTopics.contains(topic))
        throw new TopicExistsException(s"Topic '$topic' already exists.")
      val collidingTopics = allTopics.filter(Topic.hasCollision(topic, _))
      if (collidingTopics.nonEmpty) {
        throw new InvalidTopicException(s"Topic '$topic' collides with existing topics: ${collidingTopics.mkString(", ")}")
      }
    }
```

ijuma commented on core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala:
```diff
@@ -528,7 +528,7 @@ case class SetAclRequest(path: String, acl: Seq[ACL], version: Int, ctx: Option[
 }
 
-case class GetChildrenRequest(path: String, ctx: Option[Any] = None) extends AsyncRequest {
+case class GetChildrenRequest(path: String, registerWatch: Boolean = true, ctx: Option[Any] = None) extends AsyncRequest {
```
I would remove the default since it has serious implications. It should be chosen explicitly.

dajac replied:
That's a fair point. Let me remove it.

stanislavkozlovski commented:
@dajac we check for collisions because Kafka converts metric tags (topic names) dots to underscores - https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/metrics/KafkaMetricsGroup.scala#L95
Originally introduced by [this commit](https://github.com/apache/kafka/commit/1d2bd6284b06b579c901e6be8919a8a27dbe11ee) and asked [here](http://mail-archives.apache.org/mod_mbox/kafka-dev/201507.mbox/%3CCAHBV8WcMTs04xV9FjLtcRquegd2TkEPOdeuZ_O8z583pC_SkBA%40mail.gmail.com%3E)

Seems like we need to check the names otherwise we may get duplicate metric tags for some topic names

ijuma commented:
@dajac We call that method in every case, right? Not just in case the policy is defined (createTopicWithAssignment calls it as well). I agree that the cost is too high for what it's doing. Relying on the metadata cache for that use case seems reasonable (as long as we're careful about the fact that propagation is async).

In addition, have we verified that it's safe to call `getAllTopicsInCluster` without registering watches? Since watches are one shot, this could break things. cc @junrao

dajac commented:
@ijuma Yes, that method is called in all cases. My mistake. OK, I will try to address this in the another PR if you don't mind.

Regarding calling `getAllTopicsInCluster` without registering watches, I think that it is safe. There are two important places where the watch must be registered in the controller: 1) in the `initializeControllerContext` method; and 2) in the `processTopicChange` method. The former does the initial registering and the latter re-register every time the event is triggered (and processed). To my knowledge, this is sufficient. I have added the `testGetAllTopicsInClusterTriggersWatch` which mimics this and also verifies that an extra call without registering a watch does not break it.

That being said, having a second confirmation would be appreciated for sure.

dajac committed:
refactor

junrao commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -457,18 +457,25 @@ class KafkaZkClient private[zk] (zooKeeperClient: ZooKeeperClient, isSecure: Boo
+   * Gets all topics in the cluster. It does not register a watch.
+   * @return sequence of topics in the cluster.
+   */
+//  def getAllTopicsInCluster(): Set[String] = getAllTopicsInCluster(false)
```
Do we need the // ?

dajac replied:
I have removed it entirely as it is not used.

junrao commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -457,18 +457,25 @@ class KafkaZkClient private[zk] (zooKeeperClient: ZooKeeperClient, isSecure: Boo
 
   /**
    * Gets all topics in the cluster.
+   * @param registerWatch indicates if a what must be registered or not
```
what => watch?

junrao commented on core/src/test/scala/unit/kafka/zookeeper/ZooKeeperClientTest.scala:
```diff
@@ -421,13 +421,39 @@ class ZooKeeperClientTest extends ZooKeeperTestHarness {
+      CreateRequest(child1Path, Array.empty[Byte], ZooDefs.Ids.OPEN_ACL_UNSAFE.asScala, CreateMode.PERSISTENT))
+    assertEquals("Response code for create child1 should be OK", Code.OK, createResponseChild1.resultCode)
+    assertFalse("Child change notification received",
+      zNodeChildChangeHandlerCountDownLatch.await(5, TimeUnit.SECONDS))
```
This means this unit test always needs to wait for 5 secs? That's probably too long.

dajac replied:
Correct. I have reduced it to 1s.

junrao commented on core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala:
```diff
@@ -194,7 +194,56 @@ class KafkaZkClientTest extends ZooKeeperTestHarness {
+
+    zkClient.createTopicAssignment(topic1, Map.empty)
+
+    assertFalse("Received watch notification",
```
This means this unit test always needs to wait for 5 secs? That's probably too long.

dajac replied:
Correct. I have reduced it to 1s.

junrao commented on core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala:
```diff
@@ -194,7 +194,56 @@ class KafkaZkClientTest extends ZooKeeperTestHarness {
+    assertTrue(zkClient.topicExists(topic1))
+  }
+
+  def registerChildChangeHandler(count: Int): CountDownLatch = {
```
Could this method be private?

dajac replied:
Yes. Changed.

junrao reviewed (approved):
@dajac : Thanks for the PR. LGTM. Just a few minor comments below.

As for the topic name collision check, I agree that it's much cheaper if this can be done using the metadata cache and is probably good enough. That can be done in a separate PR though.

Could you also share some the performance improvement result, if any?

dajac committed:
Address Jun's comments

dajac commented:
@junrao Thank you for the review. I have addressed your comments.

Regarding the performance improvement, I have run a load test which created 5000 topics (10 requests/sec, 50 topics/request) with 4 partitions and a replication factor of 3.

Prior to this change, 5000 `ChangeTopic` events were placed in the queue. As a consequence, the request batching was degraded substantially leading to significant amount of `UpdateMetadataResponseReceived` and `LeaderAndIsrResponseReceived`. During this test, the controller was busy during 20 minutes to process the backlog of events.

With this change, 300 `ChangeTopic` events were placed in the queue. The controller took 3 minutes to process all the events and had room to process other events.

junrao commented on core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala:
```diff
@@ -194,7 +194,56 @@ class KafkaZkClientTest extends ZooKeeperTestHarness {
+    zkClient.createTopicAssignment(topic1, Map.empty)
+
+    assertFalse("Received watch notification",
+      latch.await(1, TimeUnit.SECONDS))
```
A unit test always taking 1 sec is still long. Could we reduce this to 100ms?

dajac replied:
I did few runs locally with 100ms and it seems to work. I have reduces it to 100ms.

junrao commented on core/src/test/scala/unit/kafka/zookeeper/ZooKeeperClientTest.scala:
```diff
@@ -418,14 +418,43 @@ class ZooKeeperClientTest extends ZooKeeperTestHarness {
+      CreateRequest(child1Path, Array.empty[Byte], ZooDefs.Ids.OPEN_ACL_UNSAFE.asScala, CreateMode.PERSISTENT))
     assertEquals("Response code for create child1 should be OK", Code.OK, createResponseChild1.resultCode)
-    assertTrue("Failed to receive child change notification", zNodeChildChangeHandlerCountDownLatch.await(5, TimeUnit.SECONDS))
+    assertFalse("Child change notification received",
```
A unit test always taking 1 sec is still long. Could we reduce this to 100ms?

dajac replied:
I did few runs locally with 100ms and it seems to work. I have reduces it to 100ms.

junrao reviewed (commented):
@dajac : Thanks for the updated PR. Just a couple of minor comments.

dajac committed:
Tweak timeouts

dajac commented:
@junrao I have updated the PR based on your comments. Thanks!

dajac committed:
fixup

junrao commented on core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala:
```diff
@@ -194,7 +194,56 @@ class KafkaZkClientTest extends ZooKeeperTestHarness {
+    zkClient.createTopicAssignment(topic1, Map.empty)
+
+    assertTrue("Failed to receive watch notification",
+      latch.await(100, TimeUnit.MILLISECONDS))
```
In this case, normally the test will complete well below the timeout. So, we could set the timeout to be sth more generous like 5 secs to avoid flakiness.

In the other cases where the test always wait for the timeout, then we want to set the timeout to a much smaller value to control the testing time.

dajac replied:
Right. My mistake. I put them on par while trying various timeouts in order to avoid putting a value too low. I did not pay attention that I forgot to change them back.

junrao commented on core/src/test/scala/unit/kafka/zookeeper/ZooKeeperClientTest.scala:
```diff
@@ -418,14 +418,43 @@ class ZooKeeperClientTest extends ZooKeeperTestHarness {
+      CreateRequest(child1Path, Array.empty[Byte], ZooDefs.Ids.OPEN_ACL_UNSAFE.asScala, CreateMode.PERSISTENT))
+    assertEquals("Response code for create child1 should be OK", Code.OK, createResponseChild1.resultCode)
+    assertTrue("Failed to receive child change notification",
+      zNodeChildChangeHandlerCountDownLatch.await(100, TimeUnit.MILLISECONDS))
```
This is another case where we can set a more generous timeout.

dajac replied:
Done.

junrao reviewed (commented):
@dajac : Thanks for the updated PR. A couple of more minor comments below.

dajac committed:
change timeouts

dajac commented:
@junrao Could you start the tests please?

hachikuji commented:
ok to test

hachikuji commented:
retest this please

hachikuji commented:
https://builds.apache.org/job/kafka-pr-jdk11-scala2.13/4861/

dajac commented:
@hachikuji or @junrao I think this one is in good shape now. What about merging it?

One test failed:
* kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup

I think that it is a flaky one.

dajac commented:
retest this please

dajac commented:
retest this please
